rust/connlib/tunnel/src/client.rs

```diff
@@ -233,22 +233,36 @@ pub struct AwaitingConnectionDetails {
 }
 
 impl ClientState {
-    pub(crate) fn handle_new_packet<'b>(
+    /// Attempt to handle the given packet as a DNS packet.
+    ///
+    /// Returns `Ok` if the packet is in fact a DNS query with an optional response to send back.
+    /// Returns `Err` if the packet is not a DNS query.
```

Comment:
> For non-public interfaces we should use normal comments.

Why? I find the rustdoc style quite an effective way of writing documentation.

> I don't like returning an `Error` when it's not an error.

But it is an error, the packet is not a valid DNS packet. There are other APIs like this, see for example https://docs.rs/futures/latest/futures/channel/oneshot/struct.Sender.html#method.send.

> It suggest something that implements the `Error` interface and that we could coerce using `?`

FTR: `?` has nothing to do with the `Error` trait. `?` will work on anything that implements the `Try` trait so it will also work on this API assuming that calling function returns a `Result`[^1] and its `Err` value can be constructed from this using `From`.

> I'd rather use something like `Either` or our own enum.

I can use `Either` if you feel strongly but honestly, that just makes things less clear because `Either::Left` and `Either::Right` don't have any semantic meaning when `handle_dns` has clearly two outcomes:

1. We successfully handled the packet.
2. We failed to handle the packet.

For performance reasons, we take the packet by value so we have to return it again in the error case.

I'd prefer to leave it as is for the semantic meaning that `Ok` and `Err` give to the API. In my eyes `Either` weighs both variants "equally" which is not appropriate here.

[^1]: With try_v2, this might not even be a hard-requirement but I've not looked at the implementation in detail in a while so can't say for sure.